ENH Array API support for f1_score and multilabel_confusion_matrix

OmarManzoor opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Towards #26024 

#### What does this implement/fix? Explain your changes.
- Adds array api support for f1_score and the functions related to it.
- Converts the relevant metric values to a float so that a scalar is returned. Ref: #27339

#### Any other comments?
CC: @ogrisel @betatim 

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

OmarManzoor committed:
ENH Array API support for f1_score

OmarManzoor commented on sklearn/preprocessing/_label.py:
```diff
@@ -129,7 +129,7 @@ def transform(self, y):
         """
         check_is_fitted(self)
-        y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)
+        y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)  # Pytorch breaks
```
@ogrisel @betatim I think we might need to handle LabelEncoder before we can continue with f1_score.

OmarManzoor commented on sklearn/utils/_array_api.py:
```diff
@@ -232,6 +232,14 @@ def take(self, X, indices, *, axis=0):
+    def searchsorted(self, a, v, *, side="left", sorter=None):
+        indices = numpy.searchsorted(a, v, side=side, sorter=sorter)
+        return self._namespace.asarray(indices)
+
```
For these two methods since they only seem to be missing for the numpy.array_api and probably cupy.array_api is this format fine? We might have to convert to a cupy array maybe for cupy.array_api. For the time being I could not test out with CuPY because it does not seem to be available on Mac M1.

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor committed:
Add array api support for f1_score

OmarManzoor committed:
Add changelog

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor committed:
Fix sample weights in _bincount

OmarManzoor committed:
Add some fixes

OmarManzoor committed:
Correct and add tests for nanmean

OmarManzoor commented:
@ogrisel Could you kindly have a look at this PR?

OmarManzoor commented on sklearn/metrics/_classification.py:
```diff
@@ -547,47 +555,52 @@ def multilabel_confusion_matrix(
         # Select labels:
         if not np.array_equal(labels, present_labels):
-            if np.max(labels) > np.max(present_labels):
+            if xp.max(labels) > xp.max(present_labels):
```
For the multi label case should we just leave these as np?

OmarManzoor committed:
Add options for testing with various average values

OmarManzoor committed:
Use reshape when creating arrays in micro average

ogrisel commented on sklearn/metrics/_classification.py:
```diff
@@ -547,47 +555,52 @@ def multilabel_confusion_matrix(
+        tp_bins = xp.reshape(xp.take(y_true, tp_indices, axis=0), (-1,))
         if sample_weight is not None:
-            tp_bins_weights = np.asarray(sample_weight)[tp]
+            tp_bins_weights = xp.take(xp.asarray(sample_weight), tp_indices, axis=0)
```
We might need to be specific about the device here as well.

OmarManzoor replied:
According to the ArrayAPI spec https://data-apis.org/array-api/latest/API_specification/generated/array_api.asarray.html
The dtype and device are both inferred from the initial array. Since sample_weight is already set to the corresponding device in our tests
```python
    if metric_kwargs.get("sample_weight") is not None:
        metric_kwargs["sample_weight"] = xp.asarray(
            metric_kwargs["sample_weight"], device=device
        )
```
I think that is why no error is being raised.

ogrisel commented on sklearn/metrics/_classification.py:
```diff
@@ -547,47 +555,52 @@ def multilabel_confusion_matrix(
         # labels are now from 0 to len(labels) - 1 -> use bincount
         tp = y_true == y_pred
-        tp_bins = y_true[tp]
+        tp_indices = xp.nonzero(tp)[0]
+        tp_bins = xp.reshape(xp.take(y_true, tp_indices, axis=0), (-1,))
```
I am wondering why we need this. Both boolean array indexing and `xp.(flat)nonzero` have the [data-dependent output shape warning](https://data-apis.org/array-api/latest/design_topics/data_dependent_output_shapes.html#data-dependent-output-shapes):

- https://data-apis.org/array-api/latest/API_specification/indexing.html#boolean-array-indexing
- https://data-apis.org/array-api/latest/API_specification/generated/array_api.nonzero.html#nonzero

So if boolean array indexing is going to be challenging to implement for a library, I suspect that `xp.nonzero` will have a similar problem.

So I would rather just use `tp_bins = y_true[tp]` directly and similarly for `tp_bins_weights`.

ogrisel commented on sklearn/metrics/_classification.py:
```diff
@@ -607,19 +620,19 @@ def multilabel_confusion_matrix(
-        tp = np.array(tp)
-        fp = np.array(fp)
-        fn = np.array(fn)
+        sample_weight = xp.asarray(sample_weight)
```
Here is the output of my cuda run on this PR (updated to check that boolean array indexing also works, but this should be orthogonal):

```python
$ pytest -vlx -k "array_api and f1_score" sklearn/ 

================================================================================================== test session starts ===================================================================================================
platform linux -- Python 3.10.12, pytest-7.4.2, pluggy-1.3.0
collected 34881 items / 34863 deselected / 2 skipped / 18 selected                                                                                                                                                       

sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_binary_classification_metric-numpy-None-None] PASSED                                                                      [  5%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_binary_classification_metric-array_api_strict-None-None] PASSED                                                           [ 11%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_binary_classification_metric-cupy-None-None] PASSED                                                                       [ 16%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_binary_classification_metric-cupy.array_api-None-None] PASSED                                                             [ 22%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_binary_classification_metric-torch-cpu-float64] PASSED                                                                    [ 27%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_binary_classification_metric-torch-cpu-float32] PASSED                                                                    [ 33%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_binary_classification_metric-torch-cuda-float64] PASSED                                                                   [ 38%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_binary_classification_metric-torch-cuda-float32] PASSED                                                                   [ 44%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_binary_classification_metric-torch-mps-float32] SKIPPED (Skipping MPS device test because PYTORCH_ENABLE_MPS_FALLBACK...) [ 50%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_multiclass_classification_metric-numpy-None-None] PASSED                                                                  [ 55%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_multiclass_classification_metric-array_api_strict-None-None] PASSED                                                       [ 61%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_multiclass_classification_metric-cupy-None-None] PASSED                                                                   [ 66%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_multiclass_classification_metric-cupy.array_api-None-None] PASSED                                                         [ 72%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_multiclass_classification_metric-torch-cpu-float64] PASSED                                                                [ 77%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_multiclass_classification_metric-torch-cpu-float32] PASSED                                                                [ 83%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_multiclass_classification_metric-torch-cuda-float64] PASSED                                                               [ 88%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_multiclass_classification_metric-torch-cuda-float32] PASSED                                                               [ 94%]
sklearn/metrics/tests/test_common.py::test_array_api_compliance[f1_score-check_array_api_multiclass_classification_metric-torch-mps-float32] SKIPPED (Skipping MPS device test because PYTORCH_ENABLE_MPS_FALL...) [100%]

============================================================================= 16 passed, 4 skipped, 34863 deselected, 105 warnings in 15.59s =============================================================================
```

ogrisel reviewed (commented):
Overall this looks good to me. I am surprised it works without being very specific about device and dtypes, but as long as the tests (and they do), I am happy.

OmarManzoor committed:
Add LabelEncoder and f1_score in array_api.rst

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor commented:
@ogrisel @betatim Does this look okay now?

ogrisel commented on sklearn/metrics/_classification.py:
```diff
@@ -1787,10 +1801,11 @@ def precision_recall_fscore_support(
-        tp_sum = np.array([tp_sum.sum()])
-        pred_sum = np.array([pred_sum.sum()])
-        true_sum = np.array([true_sum.sum()])
+        tp_sum = xp.reshape(xp.asarray(xp.sum(tp_sum)), (-1,))
+        pred_sum = xp.reshape(xp.asarray(xp.sum(pred_sum)), (-1,))
+        true_sum = xp.reshape(xp.asarray(xp.sum(true_sum)), (-1,))
```
Wouldn't the following work?

```suggestion
        tp_sum = xp.asarray([xp.sum(tp_sum)])
        pred_sum = xp.asarray([xp.sum(pred_sum)])
        true_sum = xp.asarray([xp.sum(true_sum)])
```

Alternatively I would prefer replacing the `(-1,)` shapes by `(1,)` to be more explicit.

OmarManzoor replied:
cupy.array_api fails with the provided suggestion. The complete stack trace is present in the details

<details>

```console
_ test_array_api_compliance[f1_score-check_array_api_multiclass_classification_metric-cupy.array_api-None-None] _

metric = <function f1_score at 0x7bbd4f374d30>
array_namespace = 'cupy.array_api', device = None, dtype_name = None
check_func = <function check_array_api_multiclass_classification_metric at 0x7bbd4f3ce3b0>

    @pytest.mark.parametrize(
        "array_namespace, device, dtype_name", yield_namespace_device_dtype_combinations()
    )
    @pytest.mark.parametrize("metric, check_func", yield_metric_checker_combinations())
    def test_array_api_compliance(metric, array_namespace, device, dtype_name, check_func):
>       check_func(metric, array_namespace, device, dtype_name)

sklearn/metrics/tests/test_common.py:1967: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sklearn/metrics/tests/test_common.py:1812: in check_array_api_multiclass_classification_metric
    check_array_api_metric(
sklearn/metrics/tests/test_common.py:1762: in check_array_api_metric
    metric_xp = metric(a_xp, b_xp, **metric_kwargs)
sklearn/utils/_param_validation.py:213: in wrapper
    return func(*args, **kwargs)
sklearn/metrics/_classification.py:1291: in f1_score
    return fbeta_score(
sklearn/utils/_param_validation.py:186: in wrapper
    return func(*args, **kwargs)
sklearn/metrics/_classification.py:1483: in fbeta_score
    _, _, f, _ = precision_recall_fscore_support(
sklearn/utils/_param_validation.py:186: in wrapper
    return func(*args, **kwargs)
sklearn/metrics/_classification.py:1806: in precision_recall_fscore_support
    tp_sum = xp.asarray([xp.sum(tp_sum)])
/opt/conda/lib/python3.10/site-packages/cupy/array_api/_creation_functions.py:83: in asarray
    res = np.asarray(obj, dtype=dtype)
/opt/conda/lib/python3.10/site-packages/cupy/_creation/from_data.py:88: in asarray
    return _core.array(a, dtype, False, order, blocking=blocking)
cupy/_core/core.pyx:2379: in cupy._core.core.array
    ???
cupy/_core/core.pyx:2402: in cupy._core.core.array
    ???
cupy/_core/core.pyx:2465: in cupy._core.core._array_from_nested_sequence
    ???
cupy/_core/core.pyx:2521: in cupy._core.core._array_from_nested_cupy_sequence
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = Array(2, dtype=int64), axis = (0,)

    def expand_dims(a, axis):
        """Expands given arrays.
    
        Args:
            a (cupy.ndarray): Array to be expanded.
            axis (int): Position where new axis is to be inserted.
    
        Returns:
            cupy.ndarray: The number of dimensions is one greater than that of
            the input array.
    
        .. seealso:: :func:`numpy.expand_dims`
    
        """
        if type(axis) not in (tuple, list):
            axis = axis,
>       return _manipulation._expand_dims(a, axis)
E       TypeError: Argument 'a' has incorrect type (expected cupy._core.core._ndarray_base, got Array)


````

</details>

I think we can go for reshape by (1,)

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -962,3 +982,16 @@ def _in1d(ar1, ar2, xp, assume_unique=False, invert=False):
         return xp.take(ret, rev_idx, axis=0)
+
+
+def _bincount(xp, array, weights=None, minlength=None):
```
```suggestion
def _bincount(xp, array, weights=None, minlength=None):
    # TODO: update if bincount is ever adopted in a future version of the standard: 
    # https://github.com/data-apis/array-api/issues/812
```

ogrisel reviewed (commented):
I do not have time to finish the review today but here is some quick feedback:

ogrisel committed:
Merge branch 'main' into f1_array_api

ogrisel commented:
I merged `main` to be able to launch the new CUDA GPU CI workflow on this PR.

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor committed:
Update: PR suggestions

OmarManzoor committed:
Use xp.reshape with (1,)

ogrisel commented on sklearn/utils/_array_api.py:
```diff
@@ -735,6 +735,26 @@ def _nanmax(X, axis=None, xp=None):
+    else:
+        mask = xp.isnan(X)
+        total = xp.sum(xp.where(mask, xp.asarray(0.0, device=device(X)), X), axis=axis)
+        count = xp.sum(
+            xp.where(
+                mask,
+                xp.asarray(0.0, device=device(X)),
+                xp.asarray(1.0, device=device(X)),
+            ),
+            axis=axis,
+        )
```
The following looks more straightforward:


```suggestion
        count = xp.sum(xp.astype(xp.logical_not(mask), X.dtype), axis=axis)
```

OmarManzoor replied:
Nice!

ogrisel reviewed (approved):
LGTM once the following is addressed:

OmarManzoor committed:
Simplify count in _nanmean

OmarManzoor committed:
Merge branch 'main' into f1_array_api

ogrisel commented:
@betatim this is ready for a second review.

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor commented:
@betatim Could you kindly review this PR?

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor committed:
Add multilabel confusion metrics as it seems to work

OmarManzoor commented:
@ogrisel @betatim Since multilabel confusion matrix seems to work fine I added that in this PR. I think it might be possible to handle precision, recall and fbeta_score in one PR as well once this is merged. I also added a generic function in the tests to get any additional params combinations.

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor committed:
Handle multi-label case

OmarManzoor committed:
Fix commented tests

OmarManzoor commented:
@betatim I added the handling for the multi-label case as well. Could you kindly have a look?

ogrisel reviewed (approved):
LGTM again with the extra commits after my last review.

ogrisel commented on doc/modules/array_api.rst:
```diff
@@ -94,6 +94,7 @@ Estimators
 - :class:`linear_model.Ridge` (with `solver="svd"`)
 - :class:`discriminant_analysis.LinearDiscriminantAnalysis` (with `solver="svd"`)
 - :class:`preprocessing.KernelCenterer`
+- :class:`preprocessing.LabelEncoder`
```
Note: this was introduced in #27381 but was missing from the changelog.

ogrisel committed:
Merge branch 'main' into f1_array_api

ogrisel commented:
Merged `main` and labeled with "CUDA CI" to check that the tests still pass with the recent CI config and dependency updates.

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor commented:
@ogrisel I think the tests are failing because in the error messages we have numpy._str values that arise because numpy 2.0 supports the array api by default. Do we want to fix such errors by turning such objects to strs before creating the error message or do we need to place a check at the place where the creation of numpy._str objects id done?

OmarManzoor committed:
Fix errors because of update in numpy

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor committed:
Minor updates

OmarManzoor committed:
Update the conversion to float

OmarManzoor committed:
Fix the doctests

OmarManzoor committed:
Fix f1 score doctests

OmarManzoor commented:
@ogrisel Could you have a look to see if this looks okay?

ogrisel commented on sklearn/metrics/_classification.py:
```diff
@@ -555,77 +564,106 @@ def multilabel_confusion_matrix(
+                device=device_,
+                axis=sum_axis,
+                sample_weight=sample_weight,
+            )
```
Is the only difference between the two branches the `device=device_` keyword argument passing?

Why not just drop the NumPy-specific branch? The array API branch should always work, no?

OmarManzoor replied:
The first branch is specifically for handling sparse arrays. That was the default before introducing the array api and thats why we have this split.

OmarManzoor replied:
The count_nonzero is the function that was already present by default and it handles sparse arrays. The _count_nonzero is our own method because the array api code is handled differently from the sparse cases.

adrinjalali replied:
but if our `_count_nonzero` handles sparse (which we can make it to do), then we don't need two code branches here.

OmarManzoor replied:
Added the conditioning in the _count_nonzero.

ogrisel reviewed (approved):
I think the `float()` calls are an improvement.

Still LGTM besides the following:

OmarManzoor commented:
@ogrisel Added the CUDA CI here to check if the tests run properly this time.

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor commented:
@betatim Do you have the time to provide the second review for this PR?

OmarManzoor commented:
@adrinjalali Could you help with reviewing this PR?

adrinjalali commented on sklearn/metrics/_classification.py:
```diff
@@ -1911,10 +1956,10 @@ def precision_recall_fscore_support(
-        recall = _nanaverage(recall, weights=weights)
-        f_score = _nanaverage(f_score, weights=weights)
+        assert average != "binary" or precision.shape[0] == 1
+        precision = float(_nanaverage(precision, weights=weights))
+        recall = float(_nanaverage(recall, weights=weights))
+        f_score = float(_nanaverage(f_score, weights=weights))
```
so we're also converting these results to python types at the same time, a link to https://github.com/scikit-learn/scikit-learn/issues/27339 would be nice in the description or the PR probably.

OmarManzoor replied:
Added in the description of the PR

adrinjalali commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -2118,3 +2148,21 @@ def test_metrics_dataframe_series(metric_name, df_lib_name):
+
+
+def _get_metric_kwargs_for_array_api_testing(metric, params):
+    metric_kwargs_combinations = [{}]
```
could you please add a docstring explaining what this does

adrinjalali commented on sklearn/utils/_array_api.py:
```diff
@@ -1055,3 +1068,18 @@ def _modify_in_place_if_numpy(xp, func, *args, out=None, **kwargs):
     return out
+
+
+def _bincount(xp, array, weights=None, minlength=None):
```
I think `xp` should be the last param like many other methods here.

OmarManzoor replied:
We could add it as the last parameter but it has to made optional then. Currently the way this function is implemented it is kind of required. Otherwise we would need to extract xp from the array once again in case it is None.

OmarManzoor commented on sklearn/metrics/_classification.py:
```diff
@@ -1325,21 +1363,21 @@ def f1_score(
     >>> y_true = [0, 1, 2, 0, 1, 2]
     >>> y_pred = [0, 2, 1, 0, 0, 1]
     >>> f1_score(y_true, y_pred, average='macro')
-    np.float64(0.26...)
+    0.26...
```
No we are converting to float ourselves when returning.

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor committed:
Add new changelog

OmarManzoor commented on sklearn/metrics/_classification.py:
```diff
@@ -520,9 +525,11 @@ def multilabel_confusion_matrix(
     y_type, y_true, y_pred = _check_targets(y_true, y_pred)
     if sample_weight is not None:
-        sample_weight = column_or_1d(sample_weight)
+        sample_weight = xp.asarray(column_or_1d(sample_weight), device=device_)
```
`column_or_1d` does not handle assigning a device currently.

adrinjalali replied:
then shouldn't it? Seems like something that we'd want to fix.

OmarManzoor commented on sklearn/metrics/_classification.py:
```diff
@@ -555,77 +564,106 @@ def multilabel_confusion_matrix(
         # Select labels:
-        if not np.array_equal(labels, present_labels):
-            if np.max(labels) > np.max(present_labels):
+        if labels.shape != present_labels.shape or xp.any(
+            xp.not_equal(labels, present_labels)
+        ):
```
I think since this can be easily handled using xp functions, we should probably not add an additional utils function.

OmarManzoor commented on sklearn/metrics/_classification.py:
```diff
@@ -1621,14 +1661,18 @@ def _check_set_wise_labels(y_true, y_pred, average, labels, pos_label):
+                    if isinstance(present_labels[0], np.generic):
+                        valid_labels = [lab.item() for lab in present_labels]
+                    else:
+                        valid_labels = present_labels
```
This test raises an error if I don't check for this. Probably has to do with numpy 2.0 and not this PR particularly.
https://github.com/scikit-learn/scikit-learn/blob/99f0f69fd72f0f1f171e465aa485589451ddb7e7/sklearn/metrics/tests/test_classification.py#L1490-L1495

```console
 ValueError: pos_label=2 is not a valid label. It should be one of [np.int64(0), np.int64(1)]
```

adrinjalali replied:
could you maybe investigate then why? this seems suspicious

OmarManzoor replied:
Basically originally the tolist() function in numpy handles this behavior by converting the inner elements using the .item() function. I added a helper in the array api to handle this behavior.

OmarManzoor committed:
Add docstring

OmarManzoor committed:
Add xp after array param in _bincount

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor committed:
Refactor based on PR suggestions

OmarManzoor committed:
Handle the _tolist method and add doc for device param

OmarManzoor committed:
Use _convert_to_numpy is _tolist

OmarManzoor commented:
@adrinjalali Could you kindly have a look at this PR now?

adrinjalali commented on sklearn/metrics/_classification.py:
```diff
@@ -578,77 +588,102 @@ def multilabel_confusion_matrix(
-                tp_bins, weights=tp_bins_weights, minlength=len(labels)
+        if tp_bins.shape[0]:
+            tp_sum = _bincount(
+                tp_bins, xp, weights=tp_bins_weights, minlength=labels.shape[0]
```
For consistency, I prefer `xp` to be the last `kwarg` like the other funcs we have.

adrinjalali commented on sklearn/metrics/_classification.py:
```diff
@@ -578,77 +588,102 @@ def multilabel_confusion_matrix(
-        tp_sum = tp_sum[indices]
-        true_sum = true_sum[indices]
-        pred_sum = pred_sum[indices]
+        indices = _searchsorted(xp, sorted_labels, labels[:n_labels])
```
same with the signature here

adrinjalali commented on sklearn/metrics/_classification.py:
```diff
@@ -578,77 +588,102 @@ def multilabel_confusion_matrix(
             y_true = y_true[:, labels[:n_labels]]
             y_pred = y_pred[:, labels[:n_labels]]
 
+        if _is_numpy_namespace(xp=xp):
+            true_and_pred = y_true.multiply(y_pred)
+        else:
+            true_and_pred = xp.multiply(y_true, y_pred)
```
why the difference?

OmarManzoor replied:
The first branch is the case where we are multiplying two sparse matrices together.

adrinjalali replied:
then we should check for sparse, not numpy namespace. This looks to me that we'd be using that branch for `np.ndarray`

OmarManzoor replied:
Done.

adrinjalali commented on sklearn/metrics/_classification.py:
```diff
@@ -578,77 +588,102 @@ def multilabel_confusion_matrix(
-            true_and_pred, axis=sum_axis, sample_weight=sample_weight
+        tp_sum = _count_nonzero(
+            true_and_pred,
+            xp=xp,
```
same with the signature / kwarg order

adrinjalali commented on sklearn/utils/extmath.py:
```diff
@@ -1228,24 +1228,24 @@ def _nanaverage(a, weights=None):
 
     if weights is None:
-        return np.nanmean(a)
+        return _nanmean(a)
```
missing a pass of `xp`

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1429,6 +1429,11 @@ def column_or_1d(y, *, dtype=None, warn=False):
        To control display of warnings.
 
+    device : device, default=None
+        `device` object (see the "Device Support" section of the array API spec).
```
should be an actual `:ref:` to render in the docs.

OmarManzoor committed:
Further PR suggestions

OmarManzoor committed:
Add sparse check specifically when multiplying

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor commented:
@adrinjalali Does this okay now?

adrinjalali commented on sklearn/metrics/_classification.py:
```diff
@@ -578,77 +588,102 @@ def multilabel_confusion_matrix(
             y_true = y_true[:, labels[:n_labels]]
             y_pred = y_pred[:, labels[:n_labels]]
 
+        if issparse(y_true) and issparse(y_pred):
```
should this be `and` or `or`?

OmarManzoor replied:
I think we can make it `or`. But not sure if this would work if y_true is sparse and y_pred is an array or the other way around.

OmarManzoor replied:
Moreover I think we convert both of these to sparse in the original code ourselves, so either would work I suppose.

adrinjalali commented on sklearn/metrics/_classification.py:
```diff
@@ -1955,10 +1993,10 @@ def precision_recall_fscore_support(
-        precision = _nanaverage(precision, weights=weights)
-        recall = _nanaverage(recall, weights=weights)
-        f_score = _nanaverage(f_score, weights=weights)
+        assert average != "binary" or precision.shape[0] == 1
```
I'm okay to leave this as is in this PR with this since it's existing code, but we really shouldn't be `assert` ing here. If this never happens, then the line shouldn't be here, if it can happen, we should raise a meaningful error.

adrinjalali reviewed (commented):
otherwise LGTM.

OmarManzoor committed:
Merge branch 'main' into f1_array_api

OmarManzoor committed:
Update 'and' to 'or' condition

adrinjalali commented on sklearn/utils/validation.py:
```diff
@@ -1429,6 +1429,12 @@ def column_or_1d(y, *, dtype=None, warn=False):
+        `device` object.
+        See the :ref:`Array API User Guide <array_api>` for more details.
+
+        .. versionadded:: 1.6
```
@glemaitre @jeremiedbb should we change this or backport?

glemaitre replied:
Let's backport since we have the experimental guardrail

adrinjalali reviewed (approved):
Other than the version question, LGTM.

glemaitre commented:
Thanks @OmarManzoor
